SPIR-V support for GLSL texture functions

jkwak-work opened the pull request:
Closes #4147

This commit implements GLSL texture functions with SPIR-V intrinsics. It also implements some of missing GLSL implementations.
 - textureProj
 - textureLod
 - texelFetchOffset
 - textureProjOffset
 - textureLodOffset
 - textureProjLod
 - textureProjLodOffset
 - textureGrad
 - textureGradOffset
 - textureProjGrad
 - textureProjGradOffset

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler1D<vector<T,N>> sampler, vec2 p)
 {
-    return texture(sampler, p.x / p.y);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

Since `textureProj` is not calling `texture()` which transitively calls `__requireComputeDerivative()` we must call `__requireComputeDerivative()` in this function to allow the implicit LOD function to work with compute shaders.

jkwak-work replied:
I am trying to figure out how to prevent a regression of cases like this where `__requireComputeDerivative()` is missing.
Looking at the test you mentioned, `tests\glsl-intrinsic\compute-derivative\intrinsic-derivative-function-in-compute.slang`, I think it should be catching those cases.
But with my change, they are passing.
I am thinking we need a new issue for this case and improve the existing test, "intrinsic-derivative-function-in-compute.slang".
What do you think?

ArielG-NV replied:
The current `intrinsic-derivative-function-in-compute.slang` test is not done 'correctly'. To test it we would need 1 test per function.

jkwak-work replied:
I found that the capability/extension is added by other functions and that might be a reason of why the existing test is passing.
I also found that we are adding more than just once and it ended up adding the same line many times.
```
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
OpCapability ComputeDerivativeGroupQuadsNV
```

jkwak-work replied:
For the repeated OpCapability, I created a new issue, #4185

jkwak-work replied:
I will create a new PR for the missing `__requireComputeDerivative()`.
I think it needs to be addressed with a new test and I want to separate it from this PR.

ArielG-NV replied:
> I found that the capability/extension is added by other functions and that might be a reason of why the existing test is passing.

yes, we need 1 test per function

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler1D<vector<T,N>> sampler, vec4 p, float bias)
 {
-    return texture(sampler, p.x / p.w, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative()
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler2D<vector<T,N>> sampler, vec3 p)
 {
-    return texture(sampler, p.xy / p.z);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative()
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler2D<vector<T,N>> sampler, vec3 p, float bias)
 {
-    return texture(sampler, p.xy / p.z, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative()
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler2D<vector<T,N>> sampler, vec4 p)
 {
-    return texture(sampler, p.xy / p.w);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative()
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
-    return texture(sampler, p.xy / p.w, bias);
+    __target_switch
+    {
+    case glsl: __intrinsic_asm "textureProj";
+    case spirv: return spirv_asm {
```
```suggestion
    __requireComputeDerivative()
    case glsl: __intrinsic_asm "textureProj";
    case spirv: return spirv_asm {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler2D<vector<T,N>> sampler, vec4 p, float bias)
 {
-    return texture(sampler, p.xy / p.w, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler3D<vector<T,N>> sampler, vec4 p)
 {
-    return texture(sampler, p.xyz / p.w);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public vector<T,4> textureProj(Sampler3D<vector<T,N>> sampler, vec4 p, float bias)
 {
-    return texture(sampler, p.xyz / p.w, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public float textureProj(sampler1DShadow sampler, vec4 p)
 {
-    return texture(sampler, p.xyz / p.w);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public float textureProj(sampler1DShadow sampler, vec4 p, float bias)
 {
-    return texture(sampler, p.xyz / p.w, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 public float textureProj(sampler2DShadow sampler, vec4 p, float bias)
 {
-    return texture(sampler, p.xyz / p.w, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler1D<vector<T,N>> sampler, vec2 p, constexpr int offset)
 {
-    return textureOffset(sampler, p.x / p.y, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler1D<vector<T,N>> sampler, vec2 p, constexpr int offset, float bias)
 {
-    return textureOffset(sampler, p.x / p.w, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler1D<vector<T,N>> sampler, vec4 p, constexpr int offset)
 {
-    return textureOffset(sampler, p.xy / p.z, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler1D<vector<T,N>> sampler, vec4 p, constexpr int offset, float bias)
 {
-    return textureOffset(sampler, p.xy / p.w, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler2D<vector<T,N>> sampler, vec3 p, constexpr ivec2 offset)
 {
-    return textureOffset(sampler, p.xyz / p.w, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler2D<vector<T,N>> sampler, vec3 p, constexpr ivec2 offset, float bias)
 {
-    return textureOffset(sampler, p.xyz / p.w, offset, bias);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler2D<vector<T,N>> sampler, vec4 p, constexpr ivec2 offset)
 {
-    return __vectorReshape<4>(sampler.SampleLevel(p, lod, offset));
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+public vector<T,4> textureProjOffset(Sampler2D<vector<T,N>> sampler, vec4 p, constexpr ivec2 offset, float bias)
 {
-    return __vectorReshape<4>(sampler.SampleLevel(p, lod, offset));
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
 {
-    // TODO: Need to apply lod
-    return sampler.SampleCmpLevelZero(p.x, p.z, offset);
+    __target_switch
+    {
```
```suggestion
    __requireComputeDerivative();
    __target_switch
    {
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
 {
-    // TODO: Need to apply lod
-    return sampler.SampleCmpLevelZero(p.xy, p.z, offset);
+    __target_switch
```
```suggestion
    __requireComputeDerivative();
    __target_switch
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
-    // TODO: Need to apply lod
-    return sampler.SampleCmpLevelZero(p.xy, p.z, offset);
-}
+    __target_switch
```
```suggestion
    __requireComputeDerivative();
    __target_switch
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+[require(glsl_hlsl_spirv, texture_shadowlod)]
+public float textureProjOffset(sampler1DShadow sampler, vec4 p, constexpr int offset, float bias)
+{
+    __target_switch
```
```suggestion
    __requireComputeDerivative();
    __target_switch
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+[require(glsl_hlsl_spirv, texture_shadowlod)]
+public float textureProjOffset(sampler2DShadow sampler, vec4 p, constexpr ivec2 offset)
+{
+    __target_switch
```
```suggestion
    __requireComputeDerivative();
    __target_switch
```

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2590,110 +2737,348 @@ public vector<T,4> texelFetchOffset(Sampler2DRect<vector<T,N>> sampler, ivec2 p,
+[require(glsl_hlsl_spirv, texture_shadowlod)]
+public float textureProjOffset(sampler2DShadow sampler, vec4 p, constexpr ivec2 offset, float bias)
+{
+    __target_switch
```
```suggestion
    __requireComputeDerivative();
    __target_switch
```

ArielG-NV commented on tests/glsl-intrinsic/intrinsic-texture.slang:
```diff
@@ -1,25 +1,15 @@
-
-//RWStructuredBuffer<float> outputBuffer;
-//layout(location = 0) out vec4 o_color;
+//TEST:SIMPLE(filecheck=HLSL): -allow-glsl -stage compute  -entry computeMain -target hlsl -DCOMPUTE
```
may be worth adding compute derivative tests for hlsl/glsl implicit LOD functions under the `-DCOMPUTE` tests.

Syntax is in `tests\glsl-intrinsic\compute-derivative\intrinsic-derivative-function-in-compute.slang`

jkwak-work replied:
Thx for the pointer. I will check it out

jkwak-work replied:
Were you referring to this test?
```
bool derivativeFunc()
{
    return true
        && dFdx(1.0f) != -1.0f
        && dFdy(1.0f) != -1.0f
        && dFdxFine(1.0f) != -1.0f
        && dFdyFine(1.0f) != -1.0f
        && dFdxCoarse(1.0f) != -1.0f
        && dFdyCoarse(1.0f) != -1.0f
        && fwidth(1.0f) != -1.0f
        && fwidthFine(1.0f) != -1.0f
        && fwidthCoarse(1.0f) != -1.0f
        ;
}
```
These are not texture functions.
If needed, they should be in a separate test file.
This test is already too big.

ArielG-NV replied:
I refer to:
```
        && int(0) == textureSize(gsampler1D, int(0))
        && vec2(0) == textureQueryLod(gsampler1D, float(0))
        && int(0) == textureQueryLevels(gsampler1D)
        && gvec4(T(0)) == texture(gsampler1D, float(0))
        && gvec4(T(0)) == texture(gsampler1D, float(0), float(0))
        && gvec4(T(0)) == textureProj(gsampler1D, vec2(0))
        && gvec4(T(0)) == textureProj(gsampler1D, vec2(0), float(0))
        && gvec4(T(0)) == textureProj(gsampler1D, vec4(0))
        && gvec4(T(0)) == textureProj(gsampler1D, vec4(0), float(0))
        && gvec4(T(0)) == textureLod(gsampler1D, float(0), float(0))
        && gvec4(T(0)) == textureOffset(gsampler1D, float(0), __LINE__)
        && gvec4(T(0)) == textureOffset(gsampler1D, float(0), __LINE__, float(0))
        && gvec4(T(0)) == texelFetch(gsampler1D, int(0), int(0))
        && gvec4(T(0)) == texelFetchOffset(gsampler1D, int(0), int(0), __LINE__)
        && gvec4(T(0)) == textureProjOffset(gsampler1D, vec2(0), __LINE__)
        && gvec4(T(0)) == textureProjOffset(gsampler1D, vec2(0), __LINE__, float(0))
        && gvec4(T(0)) == textureProjOffset(gsampler1D, vec4(0), __LINE__)
        && gvec4(T(0)) == textureProjOffset(gsampler1D, vec4(0), __LINE__,float(0))
        && gvec4(T(0)) == textureLodOffset(gsampler1D, float(0), float(0), __LINE__)
        && gvec4(T(0)) == textureProjLod(gsampler1D, vec2(0), float(0))
        && gvec4(T(0)) == textureProjLod(gsampler1D, vec4(0), float(0))
        && gvec4(T(0)) == textureProjLodOffset(gsampler1D, vec2(0), float(0), __LINE__)
        && gvec4(T(0)) == textureProjLodOffset(gsampler1D, vec4(0), float(0), __LINE__)
        && gvec4(T(0)) == textureGrad(gsampler1D, float(0), float(0), float(0))
        && gvec4(T(0)) == textureGradOffset(gsampler1D, float(0), float(0), float(0), __LINE__)
        && gvec4(T(0)) == textureProjGrad(gsampler1D, vec2(0), float(0), float(0))
        && gvec4(T(0)) == textureProjGrad(gsampler1D, vec4(0), float(0), float(0))
        && gvec4(T(0)) == textureProjGradOffset(gsampler1D, vec2(0), float(0), float(0), __LINE__)
        && gvec4(T(0)) == textureProjGradOffset(gsampler1D, vec4(0), float(0), float(0), __LINE__)
        && vec4(0) == texture1D(uniform_sampler1D, float(0))
        && vec4(0) == texture1D(uniform_sampler1D, float(0), float(0))
        && vec4(0) == texture1DProj(uniform_sampler1D, vec2(0))
        && vec4(0) == texture1DProj(uniform_sampler1D, vec2(0), float(0))
        && vec4(0) == texture1DProj(uniform_sampler1D, vec4(0))
        && vec4(0) == texture1DProj(uniform_sampler1D, vec4(0), float(0))
        && vec4(0) == texture1DLod(uniform_sampler1D, float(0), float(0))
        && vec4(0) == texture1DProjLod(uniform_sampler1D, vec2(0), float(0))
        && vec4(0) == texture1DProjLod(uniform_sampler1D, vec4(0), float(0))
```

ArielG-NV replied:
these are in the same test.
the test tests for derivatives in compute shader.

Implicit LOD functions need a derivative. NV hardware has a method to get a derivative from NV hardware

jkwak-work replied:
Right.
Since this test is already too big, I will create a new test for the `__requireComputeDerivative()`; or I may modify the existing one, `intrinsic-derivative-function-in-compute.slang`.

ArielG-NV replied:
I think modify the existing one

ArielG-NV reviewed (commented):
Looks mostly good, only suggestions related to compute shader derivatives

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2231,107 +2231,239 @@ __generic<T:__BuiltinFloatingPointType, let N:int>
 [require(glsl_hlsl_spirv, texture_sm_4_1_fragment)]
```
In general for all texture functions, I wonder if we should use C++  code gen (`${{{{ }}}}`) for these texture functions since they are repetitive. May be easier to iterate on them?

jkwak-work replied:
I think that is a good idea.
But the `default` part differs on all cases.
I can make it working but I am afraid that it may make things harder to read.

csyonghe reviewed (approved):
Looks good!

jkwak-work commented:
I am sorry for the big change in the test case.
It is a lot but not hard to read.
It helped me to find a few bugs and very helpful.

I changed my mind about `__requireComputeDerivative()` and I added it to all cases in this PR.
Because I shouldn't make the build broken.

ArielG-NV commented on source/slang/glsl.meta.slang:
```diff
@@ -2368,24 +2592,60 @@ public vector<T,4> textureLod(__TextureImpl<
 {
-    // TODO: Need to apply lod
-    return sampler.SampleCmp(p.xy, p.z);
+    __requireComputeDerivative();
```
OpImageSampleDrefExplicitLod does not need `requireComputeDerivative`.

It is only needed for implicitLod functions (the derivative is used to figure out a Lod).

jkwak-work commented:
Hmm. One of tests is failing.
I am looking into it now.
```
tests/hlsl-intrinsic/texture/texture-intrinsics.slang
```
